core/src/main/java/com/segment/analytics/kotlin/core/Telemetry.kt

```diff
@@ -0,0 +1,271 @@
+                } catch (e: CancellationException) {
+                    // last flush before shutdown
+                    flush()
+                }
+            }
+        }
+    }
+
+    fun reset()
+    {
+        telemetryDispatcher?.close()
+        telemetryScope = null
+        telemetryDispatcher = null
+        resetQueue()
+        seenErrors.clear()
+        started = false
+        rateLimitEndTime = 0
+    }
+
+    fun increment(metric: String, tags: Map<String, String>) {
+        if (!enable || sampleRate == 0.0) return
+        if (!metric.startsWith("analytics_mobile.")) return
+        if (tags.isEmpty()) return
+        if (Math.random() > sampleRate) return
+        if (queue.size >= maxQueueSize) return
+
+        addRemoteMetric(metric, tags, value = (1.0 / sampleRate).roundToInt())
+    }
+
+    fun error(metric:String, tags: Map<String, String>, log: String) {
+        if (!enable || sampleRate == 0.0) return
+        if (!metric.startsWith("analytics_mobile.")) return
+        if (tags.isEmpty()) return
+        if (queue.size >= maxQueueSize) return
+
+        var filteredTags = tags
+        if (!sendWriteKeyOnError) filteredTags = tags.filterKeys { it.lowercase() != "writekey" }
+        var logData: String? = null
+        if (sendErrorLogData) {
+            logData = if (log.length > errorLogSizeMax) {
+                log.substring(0, errorLogSizeMax)
+            } else {
+                log
+            }
+        }
+
+        val errorKey = tags["error"]
+        if (errorKey != null) {
+            if (seenErrors.containsKey(errorKey)) {
+                seenErrors[errorKey] = seenErrors[errorKey]!! + 1
+                if (Math.random() > sampleRate) return
+                // Send how many we've seen after the first since we know for sure.
+                addRemoteMetric(metric, filteredTags, log=logData, value = seenErrors[errorKey]!!)
+                seenErrors[errorKey] = 0
+            } else {
+                addRemoteMetric(metric, filteredTags, log=logData)
+                flush()
+                seenErrors[errorKey] = 0 // Zero because it's already been sent.
+            }
+        }
+        else {
+            addRemoteMetric(metric, filteredTags, log=logData)
+        }
+    }
+
+    @Synchronized
+    fun flush() {
+        if (!enable || queue.isEmpty()) return
+        if (rateLimitEndTime > (System.currentTimeMillis() / 1000).toInt()) {
+            return
+        }
+        rateLimitEndTime = 0
+
+        try {
+            send()
+            queueBytes = 0
+        } catch (error: Throwable) {
+            errorHandler?.invoke(error)
+            sampleRate = 0.0
+        }
+    }
+
+    private fun send() {
+        try {
+            // Json.encodeToString by default does not include default values
+            //  We're using this to leave off the 'log' parameter if unset.
+            val payload = Json.encodeToString(mapOf("series" to queue))
+            resetQueue()
```

Comment:
might have concurrent issue (the flush thread calls this function while the main thread adds an item to the queue). we should snapshot the queue first. in addition, `mutableListOf` will throw concurrent modification exception, should change the type of the queue to a copy on write array, or change it to a real queue that we can dequeue element at the snapshotted size